python/lsst/daf/butler/script/queryCollections.py

```diff
@@ -144,51 +187,127 @@ def _getTree(
+            collection_table = hstack([collection_table, dataset_types_table]).filled("")
+        for row in collection_table:
+            table.add_row(row)
+
         if inverse:
             assert info.parents is not None  # For mypy.
             for pname in sorted(info.parents):
-                pinfo = butler.collections.get_info(pname, include_parents=inverse)
+                pinfo = butler.collections.get_info(
+                    pname, include_parents=inverse, include_summary=show_dataset_types
+                )
                 addCollection(pinfo, level + 1)
         else:
             if info.type == CollectionType.CHAINED:
                 for name in info.children:
-                    cinfo = butler.collections.get_info(name)
+                    cinfo = butler.collections.get_info(name, include_summary=show_dataset_types)
                     addCollection(cinfo, level + 1)
 
     collections = butler.collections.query_info(
-        glob or "*", collection_types=frozenset(collection_type), include_parents=inverse
+        glob or "*",
+        collection_types=frozenset(collection_type),
+        include_parents=inverse,
+        include_summary=show_dataset_types,
     )
     for collection in sorted(collections):
         addCollection(collection)
     return table
 
 
 def _getList(
-    repo: str, glob: Iterable[str], collection_type: Iterable[CollectionType], flatten_chains: bool
+    repo: str,
+    glob: Iterable[str],
+    collection_type: Iterable[CollectionType],
+    flatten_chains: bool,
+    show_dataset_types: bool = False,
+    exclude_dataset_types: Iterable[str] = [],
 ) -> Table:
     """Return collection results as a table representing a flat list of
     collections.
+
+    Parameters
+    ----------
+    repo : `str`
+        Butler repository location.
+    glob : `collections.abc.Iterable` of `str`
+        Wildcards to pass to ``queryCollections``.
+    collection_type
+        Same as `queryCollections`
+    flatten_chains : `bool`
+        If True, flatten the tree of CHAINED datasets.
+    show_dataset_types : `bool`
+        If True, also show the dataset types present within each collection.
+    exclude_dataset_types : `~collections.abc.Iterable` [ `str` ]
+        A glob-style comma-separated list of dataset types to exclude.
+        Only has an effect if `show_dataset_types` is True.
+
+    Returns
+    -------
+    collections : `astropy.table.Table`
+        Same as `queryCollections`
     """
+    table = Table(
+        names=("Name", "Type"),
+        dtype=(str, str),
+    )
+    if show_dataset_types:
+        table.add_column(Column(name="Dataset Types", dtype=str))
     butler = Butler.from_config(repo)
+
+    def addCollection(info: CollectionInfo) -> None:
+        collection_table = Table([[info.name], [info.type.name]], names=["Name", "Type"])
+        if show_dataset_types:
+            dataset_types = [""] if not info.dataset_types else info.dataset_types
```

Comment:
Since this identical code is used in three places (the filtering of dataset_types) please put it in a helper function.

Reply:
Added a `_parseDatasetTypes` helper function to perform this task.